skunkworks_crow/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -94,26 +103,60 @@ protected Result onRunJob(@NonNull Params params) {
         ((Share) getContext().getApplicationContext()).getAppComponent().inject(this);
 
         initJob(params);
-        rxEventBus.post(receiveForms());
 
         return null;
     }
 
     private void initJob(Params params) {
-        ip = params.getExtras().getString(IP, "");
-        port = params.getExtras().getInt(PORT, -1);
         sbResult = new StringBuilder();
+        boolean isBluetooth = params.getExtras().getBoolean("isBluetooth", true);
+        if (!isBluetooth) {
+            ip = params.getExtras().getString(IP, "");
+            port = params.getExtras().getInt(PORT, -1);
+        } else {
+            targetMacAddress = params.getExtras().getString("mac", null);
+        }
+
+        setupDataStreamsAndReceive(isBluetooth, targetMacAddress);
     }
 
-    private DownloadEvent receiveForms() {
-        Timber.d("Socket " + ip + " " + port);
+    private void setupDataStreamsAndReceive(boolean isBluetooth, String macAddress) {
+        try {
+            Timber.d("Waiting for receiver");
+            if (isBluetooth) {
+
+                BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
+                BluetoothDevice bluetoothDevice = bluetoothAdapter.getRemoteDevice(macAddress);
+                BluetoothSocket bluetoothSocket = bluetoothDevice.createRfcommSocketToServiceRecord(SPP_UUID);
+                if (!bluetoothSocket.isConnected()) {
+                    bluetoothSocket.connect();
```

Comment:
The thread is blocked here, and it will try to connect to the bluetooth device, if its not able to connect then it will through the `IOException` so we should fail the form transfer if exception occurs.

Reply:
So you mean we should call `cancel()` to cancel the job here? @lakshyagupta21

Reply:
No, I'm trying to say if it's not able to connect to the socket then it will through the IOException, in that case, we should stop the form transfer and for that, we need to cancel/stop the scheduled job.

Reply:
sure.